sympy/polys/tests/test_hypothesis.py

```diff
@@ -4,40 +4,33 @@
 from sympy.polys.polytools import Poly
 
 
-@given(
-    coefficients1=st.lists(st.integers()),
-    coefficients2=st.lists(st.integers()),
-    coefficients3=st.lists(st.integers()),
-)
-def test_gcd_hypothesis(coefficients1, coefficients2, coefficients3):
-    f = Poly(coefficients1, x, domain="ZZ")
-    g = Poly(coefficients2, x, domain="ZZ")
-    r = Poly(coefficients3, x, domain="ZZ")
+def polys(*, nonzero=False, domain="ZZ"):
+    # This is a simple strategy, but sufficient the tests below
+    elems = {"ZZ": st.integers(), "QQ": st.fractions()}
+    coeff_st = st.lists(elems[domain])
+    if nonzero:
+        coeff_st = coeff_st.filter(any)
+    return st.builds(Poly, coeff_st, st.just(x), domain=st.just(domain))
 
+
+@given(f=polys(), g=polys(), r=polys())
+def test_gcd_hypothesis(f, g, r):
     gcd_1 = f.gcd(g)
     gcd_2 = g.gcd(f)
-
     assert gcd_1 == gcd_2
 
     # multiply by r
     gcd_3 = g.gcd(f + r * g)
-
     assert gcd_1 == gcd_3
 
 
-@given(
-    coefficients1=st.lists(st.integers()),
-    coefficients2=st.lists(st.integers()).filter(lambda x: any(x)),
-)
-def test_poly_hypothesis(coefficients1, coefficients2):
-    # Integer case
-    f_z = Poly(coefficients1, x, domain="ZZ")
-    g_z = Poly(coefficients2, x, domain="ZZ")
+@given(f_z=polys(), g_z=polys(nonzero=True))
+def test_poly_hypothesis_integers(f_z, g_z):
     remainder_z = f_z.rem(g_z)
     assert g_z.degree() >= remainder_z.degree() or remainder_z.degree() == 0
```

Comment:
This is quote weak as a test invariant. We can be a lot more specific here. The Euclidean division of univariate polynomials `f.div(g)` should be polynomials `a` and `b` such that `f = a*g + b` and `deg(b) < `deg(g)` (strict inequality). These two invariants are sufficient to uniquely define `a` and `b` so checking them ensures that the result is guaranteed correct provided that arithmetic and degree are handled correctly.

Separately we can check that `f.div(g) == (f.quo(g), f.rem(g))`.

Also if `f.rem(g) == 0` then `f.exquo(g) == a`. Otherwise `f.exquo(g)` should raise `ExactQuotientFailed`.

Reply:
Actually what I said is true for univariate polynomials over a field (i.e. QQ rather than ZZ). I'm not sure what the invariants for `quo` etc are supposed to be for ZZ but we have `pquo`, `pdiv`, etc which are for pseudodivision of polynomials over a ring and have similar invariants but with the leading coefficient involved:
```
In [1]: R = ZZ[x]

In [2]: p1 = R(2*x**2 + 3*x + 1)

In [3]: p2 = R(3*x + 1)

In [4]: a, b = p1.pdiv(p2)

In [5]: a
Out[5]: 6*x + 7

In [6]: b
Out[6]: 2

In [7]: a*p2 + b
Out[7]: 18*x**2 + 27*x + 9

In [9]: p2.LC**2 * p1
Out[9]: 18*x**2 + 27*x + 9

In [10]: p2.LC**2 * p1 == a*p2 + b  # (1)
Out[10]: True

In [11]: b.degree()
Out[11]: 0

In [12]: b.degree() < p2.degree()  # (2)
Out[12]: True
```
Again these invariants (1) and (2) are sufficient to uniquely define a and b.

The `pquo` etc invariants apply to polynomials over a ring which means that they also apply to multivariate polynomials since any multivariate polynomial can be thought of recursively as a univariate polynomial whose coefficients are polynomials.

For multivariate polynomials we also have the concept of division by multiple divisors:
```
In [19]: R = QQ[x,y,z]

In [20]: p1 = R(2*x**2*y + 3*x + z**2)

In [21]: p2 = R(3*x + y)

In [22]: p3 = R(z**2 - 2)

In [23]: p1.div([p2, p3])
Out[23]: ([2/3*x*y - 2/9*y**2 + 1, 1], 2/9*y**3 - y + 2)

In [24]: [q2, q3], r = p1.div([p2, p3])

In [25]: p1 == q2*p2 + q3*p3 + r
Out[25]: True
```
Here the result is not necesssarily unique but it is unique if the divisors are a reduced Groebner basis and in that case the invariant is that `r` is reduced wrt the basis and so its leading monomial is less than that of each divisor wrt the chosen monomial ordering.